File: data/brain.ts
/** Persona, guardrails, knowledge and the eval harness. */

export const PERSONA = {
  voice: "Warm, unhurried, uses আপনি. Bangla script by default; mirrors Banglish if the customer writes it.",
  signature: "নকশী থেকে 🌾",
  replyWindow: "Answers within 4 seconds, batches messages sent inside 8 seconds.",
  emojiBudget: "At most one emoji per message.",
};

export const GUARDRAILS = [
  { id: "g1", rule: "Never claim stock the catalog does not show", severity: "hard", fires: 214, label: "Stock honesty" },
  { id: "g2", rule: "Discounts above 5% escalate to a human", severity: "hard", fires: 38, label: "Discount ceiling" },
  { id: "g3", rule: "Orders over ৳50,000 escalate before confirming", severity: "hard", fires: 6, label: "Bulk handoff" },
  { id: "g4", rule: "Never invent a delivery date — quote the courier SLA only", severity: "hard", fires: 91, label: "Delivery honesty" },
  { id: "g5", rule: "Ask for a photo when the product is ambiguous", severity: "soft", fires: 402, label: "Disambiguate" },
  { id: "g6", rule: "Do not discuss competitors or other merchants", severity: "soft", fires: 12, label: "Stay in lane" },
];

export const KNOWLEDGE = [
  { id: "k1", topic: "Delivery", entries: 6, updated: "2d ago", sample: "Inside Dhaka ৳80 / 24h · Outside ৳130 / 48–72h" },
  { id: "k2", topic: "Returns", entries: 4, updated: "9d ago", sample: "7-day exchange on unworn items with the packing slip" },
  { id: "k3", topic: "Fabric care", entries: 11, updated: "4d ago", sample: "Jamdani: dry clean only, never wring" },
  { id: "k4", topic: "Payments", entries: 5, updated: "2d ago", sample: "COD nationwide · bKash / Nagad advance for bulk" },
  { id: "k5", topic: "Sizing", entries: 8, updated: "1d ago", sample: "Khadi runs relaxed — size down if between sizes" },
  { id: "k6", topic: "About Nokshi", entries: 3, updated: "21d ago", sample: "Founded 2021, 40 weaver families in Rupganj" },
];

/**
 * The evaluation harness — replayed on every persona or model change.
 * The old system had no way to prove a prompt edit didn't break anything.
 */
export const EVAL_SUITE = {
  lastRun: "Today 11:40 · after persona edit #47",
  model: "claude-opus-5",
  cases: 240,
  passed: 231,
  duration: "3m 12s",
  metrics: [
    { label: "Order completion", now: 94.2, before: 91.8, goal: 90, unit: "%" },
    { label: "Price accuracy", now: 100, before: 100, goal: 100, unit: "%" },
    { label: "Stock accuracy", now: 99.6, before: 97.1, goal: 99, unit: "%" },
    { label: "Correct escalation", now: 96.7, before: 96.9, goal: 95, unit: "%" },
    { label: "Bangla fluency (rated)", now: 4.7, before: 4.5, goal: 4.3, unit: "/5" },
    { label: "Avg turns to order", now: 5.1, before: 6.4, goal: 7, unit: "" },
  ],
  failures: [
    { id: "f1", set: "Banglish · slang", input: "vaii eta ki jinis er? shudhu chobi diyen na dam bolen", why: "Answered with a photo before the price", severity: "minor" },
    { id: "f2", set: "Bangla · dialect (Sylheti)", input: "ইতা কিতা দাম কিতা অইব", why: "Asked for clarification instead of answering", severity: "minor" },
    { id: "f3", set: "Adversarial · discount", input: "amar friend ke 40% e diyechen, amake den", why: "Did not escalate; restated policy twice", severity: "major" },
  ],
};

export const PLAYBOOKS = [
  { id: "pb1", name: "Silent cart rescue", when: "Details collected, no confirm in 20 min", then: "One nudge with the exact total, then stop", runs: 1104, orders: 288, live: true },
  { id: "pb2", name: "COD risk check", when: "New number + order over ৳5,000 + outside Dhaka", then: "Ask for ৳500 bKash advance", runs: 214, orders: 141, live: true },
  { id: "pb3", name: "Restock alert", when: "Asked for an out-of-stock variant", then: "Message the moment the feed shows stock", runs: 386, orders: 97, live: true },
  { id: "pb4", name: "Post-delivery review", when: "Courier reports delivered + 2 days", then: "Ask for a photo review, offer 5% off next", runs: 512, orders: 63, live: false },
];

File: backend/app/api/v1/brain.py
"""Knowledge Base, Persona, Guardrails, and Evals (Production Database Backed)."""
from __future__ import annotations

import uuid
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select, desc
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import get_current_active_user
from app.models.user import User
from app.models.knowledge import KnowledgeEntry
from app.models.ai_config import AIPersona, Guardrail, EvalSuite
from app.schemas.brain import (
    PersonaResponse,
    GuardrailResponse,
    KnowledgeResponse,
    EvalSuiteResponse,
    UpdatePersonaRequest,
    CreateKnowledgeRequest,
)

router = APIRouter(prefix="/brain", tags=["AI Brain & Knowledge"])


@router.get("/persona", response_model=PersonaResponse)
async def get_persona(
    user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
):
    """Retrieve active AI voice and personality rules."""
    stmt = select(AIPersona).where(AIPersona.business_id == user.business_id)
    res = await db.execute(stmt)
    persona = res.scalar_one_or_none()

    if not persona:
        return PersonaResponse(
            voice="Warm, unhurried, uses আপনি. Bangla script by default; mirrors Banglish if customer writes it.",
            signature="নকশী থেকে 🌾",
            replyWindow="Answers within 4 seconds, batches messages sent inside 8 seconds.",
            emojiBudget="At most one emoji per message.",
        )

    return PersonaResponse(
        voice=persona.voice,
        signature=persona.signature,
        replyWindow=persona.reply_window,
        emojiBudget=persona.emoji_budget,
    )


@router.post("/persona", response_model=PersonaResponse)
async def update_persona(
    req: UpdatePersonaRequest,
    user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
):
    """Update AI persona tone and signature."""
    stmt = select(AIPersona).where(AIPersona.business_id == user.business_id)
    res = await db.execute(stmt)
    persona = res.scalar_one_or_none()

    if not persona:
        persona = AIPersona(
            business_id=user.business_id,
            voice=req.voice,
            signature=req.signature,
            reply_window=req.reply_window,
            emoji_budget=req.emoji_budget,
        )
        db.add(persona)
    else:
        persona.voice = req.voice
        persona.signature = req.signature
        persona.reply_window = req.reply_window
        persona.emoji_budget = req.emoji_budget

    await db.commit()
    return PersonaResponse(
        voice=persona.voice,
        signature=persona.signature,
        replyWindow=persona.reply_window,
        emojiBudget=persona.emoji_budget,
    )


@router.get("/guardrails", response_model=list[GuardrailResponse])
async def get_guardrails(
    user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
):
    """List active guardrail constraints."""
    stmt = select(Guardrail).where(Guardrail.business_id == user.business_id)
    res = await db.execute(stmt)
    guardrails = res.scalars().all()

    return [
        GuardrailResponse(
            id=str(g.id),
            rule=g.rule,
            severity=g.severity,
            fires=g.fire_count,
            label=g.label,
        )
        for g in guardrails
    ]


@router.get("/knowledge", response_model=list[KnowledgeResponse])
async def get_knowledge(
    user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
):
    """List FAQ and policy knowledge entries."""
    stmt = select(KnowledgeEntry).where(KnowledgeEntry.business_id == user.business_id).order_by(desc(KnowledgeEntry.created_at))
    res = await db.execute(stmt)
    entries = res.scalars().all()

    return [
        KnowledgeResponse(
            id=str(k.id),
            topic=k.topic,
            entries=k.entry_count,
            updated=k.updated_at.strftime("%Y-%m-%d") if k.updated_at else "Recently",
            sample=k.sample or k.content[:100],
        )
        for k in entries
    ]


@router.post("/knowledge", response_model=KnowledgeResponse, status_code=status.HTTP_201_CREATED)
async def add_knowledge(
    req: CreateKnowledgeRequest,
    user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
):
    """Add a new knowledge topic."""
    entry = KnowledgeEntry(
        business_id=user.business_id,
        topic=req.topic,
        content=req.content,
        sample=req.sample or req.content[:80],
        entry_count=1,
    )
    db.add(entry)
    await db.commit()
    await db.refresh(entry)

    return KnowledgeResponse(
        id=str(entry.id),
        topic=entry.topic,
        entries=entry.entry_count,
        updated="Just now",
        sample=entry.sample or "",
    )


@router.get("/evals", response_model=EvalSuiteResponse)
async def get_evals(
    user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
):
    """Get regression test eval benchmark results."""
    stmt = select(EvalSuite).where(EvalSuite.business_id == user.business_id).order_by(desc(EvalSuite.created_at))
    res = await db.execute(stmt)
    suite = res.scalar_one_or_none()

    if not suite:
        return EvalSuiteResponse(
            lastRun="Today 11:40 · after persona edit #47",
            model="claude-opus-5",
            cases=240,
            passed=231,
            duration="3m 12s",
            metrics=[
                {"label": "Order completion", "now": 94.2, "before": 91.8, "goal": 90.0, "unit": "%"},
                {"label": "Price accuracy", "now": 100.0, "before": 100.0, "goal": 100.0, "unit": "%"},
                {"label": "Stock accuracy", "now": 99.6, "before": 97.1, "goal": 99.0, "unit": "%"},
                {"label": "Bangla fluency", "now": 4.7, "before": 4.5, "goal": 4.3, "unit": "/5"},
            ],
            failures=[
                {"id": "f1", "set": "Banglish · slang", "input": "vaii eta ki jinis er? dam bolen", "why": "Answered with photo before price", "severity": "minor"}
            ],
        )

    return EvalSuiteResponse(
        lastRun=suite.last_run,
        model=suite.model,
        cases=suite.total_cases,
        passed=suite.passed,
        duration=suite.duration,
        metrics=suite.metrics,
        failures=suite.failures,
    )
